index.mjs

```diff
@@ -0,0 +1,437 @@
+//. the final value, discarding the final state.
+//.
+//. ```js
+//.  evalState () (Z.of (State, 1)); // 1
+//. ```
+
+export function evalState(state) {
+  return function(m) {
+    return m.run (state).value;
+  };
+}
+
+//# execState :: s -> State s a -> s
+//.
+//. Evaluate a state computation with the given initial state and return
+//. the final state, discarding the final value.
+//.
+//. ```js
+//.  execState () (put (1)); // 1
+//. ```
+
+export function execState(state) {
+  return function(m) {
+    return m.run (state).state;
+  };
+}
+
+function next(v) { return {done: false, value: v}; }
+function done(v) { return {done: true, value: v}; }
+
+//# fantasy-land/chainRec :: ((a -> c, b -> c, v) -> State s c, v) -> State s b
+//.
+//. ```js
+//.   Z.chainRec (
+//.     State,
+//.     (next, done, v) => Z.of (State, v > 10 ? done (v) : next (v + 1)),
+//.     1
+//.   ).run (null); // {state: null, value: 11}
+//. ```
+
+State['fantasy-land/chainRec'] = function chainRec(f, v) {
+  return new State (function(state) {
+    var r = {state: state, value: next (v)};
+    while (!r.value.done) {
+      r = f (next, done, r.value.value).run (r.state);
+    }
+    return {state: r.state, value: r.value.value};
+  });
+};
+
+
+//# fantasy-land/chain :: State s a ~> (a -> State s b) -> State s b
+//.
+//. Replace the State with a new one based on the result value of
+//. the previous computation.
+//.
+//. ```js
+//.   Z.chain (
+//.     v => Z.of (State, v + 1),
+//.     Z.of (State, 1)
+//.   )
+//.   .run (null); // {state: null, value: 2}
+//. ```
+
+State.prototype['fantasy-land/chain'] = function chain(f) {
+  var self = this;
+  return new State (function(s) {
+    var r = self.run (s);
+    return f (r.value).run (r.state);
+  });
+};
+
+//# fantasy-land/map :: State s a ~> (a -> b) -> State s b
+//.
+//. Map an old value to a new one. The old value is thrown away.
+//.
+//. ```js
+//.   Z.map (
+//.     x => x + 1,
+//.     Z.of (State, 1)
+//.   )
+//.   .run (null); // {state: null, value: 2}
+//. ```
+
+State.prototype['fantasy-land/map'] = function map(f) {
+  return this['fantasy-land/chain'] (compose (State['fantasy-land/of'], f));
+};
+
+//# fantasy-land/ap :: State s a ~> State s (a -> b) -> State s b
+//.
+//. Combine two State values by applying the value of one over the value
+//. within the other. The state of the applied is thrown.
+//.
+//. ```js
+//.   Z.ap (
+//.     Z.of (State, x => x + 1),
+//.     Z.of (State, 1)
+//.   )
+//.   .run (null); // {state: null, value: 2}
+//. ```
+
+State.prototype['fantasy-land/ap'] = function ap(a) {
+  return this['fantasy-land/map'] (evalState () (a));
+};
+
+//# StateT :: Monad m => m -> StateT s m a
+//.
+//. A state monad parametrised by the type m of the state to carry.
+//.
+
+export function StateT(M) {
+
+  function StateT(run) {
+    this.run = run;
+  }
+
+  //. ## StateT
+  //.
+  //# fantasy-land/of :: Monad m => a -> StateT s m a
```

Comment:
This documentation is going to conflict with the regular one. You should probably move this documentation, and all below, to the documentation for `StateT` itself, it *is* the return value of that function, after all.

Reply:
How can I document the `StateT` properties keeping the description and examples?

Reply:
We might be able to denote them as `StateT(m).fantasy-land/of :: Monad m => StateT s m a`, we're just inventing syntax to do it. I need to think about a better way, though.